### PR TITLE
S01-12: Implement image gradient computation

### PR DESCRIPTION
**Pull request by sarat12**

### CODE REVIEW

**Comment by lsem** [comment]
Verfy good commit, just see two typos so far..

File: SeamCarve.Back/src/ImageProcesing/Filters/SobelEdgeDetector.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        private byte _lowThreshold, _highThreshold, _cernelDiadonals, _cernelParalels;
+        private double _multiplier;
+
+        public SobelEdgeDetector()
+        {
+            SetDefaults();
+        }
+
+        public SobelEdgeDetector(byte lowThreshold, byte highThreshold, byte cernelDiadonals, byte cernelParalels, double multiplier)
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/Filters/SobelEdgeDetector.cs]: Why do you need to pass this kernel stuff as parameters?

Comment (sarat12) [inline SeamCarve.Back/src/ImageProcesing/Filters/SobelEdgeDetector.cs]: it could be very useful to be able то change them. For example Scharr experienced better results by setting those to 10 and 3 instead of 1 and 2 (from wiki).

File: SeamCarve.Back/src/ImageProcesing/ImageProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        };
+
+        private static BmpData _originalBmpData;
+
+        public static Bitmap Process(List<IFilter> filters, Bitmap inputBmp)
+        {
+            Debug.Assert(filters != null);
+
+            if (!SupportedPixelFormats.Contains(inputBmp.PixelFormat))
+                throw new UnsupportedPixelFormatException("Image has unsupported pixel format.");
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/ImageProcessor.cs]: As a receiver of such message, I would like to see what was expected and what was actual. Current message is not very useful.

File: SeamCarve.Back/src/ImageProcesing/ImageProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+            _originalBmpData = ImageUtility.BmpDataFromBitmap(inputBmp);
+
+#if DEBUG
+            var sw = new Stopwatch();
+            sw.Start();
+#endif
+            var result = ImageUtility.ImageIntensityMask(_originalBmpData);
+
+#if DEBUG
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/ImageProcessor.cs]: Are you still measuring performance in debug mode?

Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/ImageProcessor.cs]: You can use you own application specific symbol like #define TRACE_PERFORMANCE in Release mode instead. Please see:
https://msdn.microsoft.com/en-us/library/yt3yck0x.aspx

File: SeamCarve.Back/src/ImageProcesing/ImageUtility.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            var inBytes = imageBmpData.Bytes;
+            int inWidth = imageBmpData.Width, 
+                inHeight = imageBmpData.Height;
+            var pixSize = imageBmpData.PixelSize;
+
+            var outBytes = new byte[inWidth * inHeight];
+
+            int j = 0;
+
+            for (int i = 0; i < inBytes.Length; i += pixSize)
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/ImageUtility.cs]: technically it worth to have something like assert(pixelsize >= 3). Understand why?

File: SeamCarve.Back/src/ImageProcesing/ImageUtility.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                inHeight = intensityMask.Height;
+
+            var outBytes = new byte[inWidth * inHeight * pixelSize];
+
+            int j = 0;
+
+            for (int i = 0; i < outBytes.Length; i += pixelSize)
+            {
+                outBytes[i] = outBytes[i + 1] = outBytes[i + 2] = inBytes[j++];
+                outBytes[i + 3] = byte.MaxValue;
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/ImageUtility.cs]: You have assumption that pixel has 4 bytes, but from interface it looks like client can specify arbitary pixelsize.
You probably support only 4 bytes pixels here and in this case you should raise exceptions when something else passed.

File: SeamCarve.Back/src/ImageProcesing/Models/BmpData.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            Width = width;
+            Height = height;
+            PixelSize = pixelSize;
+        }
+
+        /// <summary>
+        /// Creates copy of BmpData instance with empty bytes
+        /// </summary>
+        /// <param name="source"></param>
+        public BmpData(BmpData source)
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/Models/BmpData.cs]: a bit confusing semantics to create a copy which is not a copy.
I would rather suggest creating static method called something like BmpData CreateShallowCopyFrom(BmpData) with the same implementation .

File: SeamCarve.Back/src/Web/Controllers/ImagesController.cs [hunk 7]
```diff
@@ -93,6 +93,23 @@ public IHttpActionResult DeleteImage(Guid id)
             return Ok("Successfully deleted");
         }
 
+        [HttpGet, Route("images/{id:guid}/energy")]
+        [ImagesApiExceptionFilter]
+        public HttpResponseMessage GetImageEnergy(Guid id, byte lowThreshold, byte highThreshold, byte kernelDiadonals, byte kernelParalels, double multiplier)
```
Comment (lsem) [inline SeamCarve.Back/src/Web/Controllers/ImagesController.cs]: is it possible to have some default one? so that I can some solution that just work? Because currently, clients basically cannot use it from PostMan.

File: SeamCarve.Back/src/tests/ApiTests/WebApiTests.cs [hunk 8]
```diff
@@ -86,6 +88,93 @@ public void UploadedAndDownloadedImagesDataShouldBeEqual()
...
+
+            var sw = new Stopwatch();
+
+            sw.Start();
+
+            SeamCarveClient.GetImageEnergyData(uploadedImageId, @params);
+
+            sw.Stop();
+
+            Assert.Greater(sw.ElapsedMilliseconds, 29);
```
Comment (lsem) [inline SeamCarve.Back/src/tests/ApiTests/WebApiTests.cs]: This kind of performance tests will work only on you machine.
Alternative approach is to track only changes. Run tests, write results to file somewhere if it not exist yet. If exists, read value, run tests, compare value. If change in performance is significant, it should be treated as regression and test becomes failed.

File: SeamCarve.Back/src/tests/ApiTests/WebApiTests.cs [hunk 9]
```diff
@@ -187,6 +276,25 @@ public void ShouldReturn404IfRequestedImageDataNotFound()
         }
 
         [Test]
+        public void ShouldReturn404IfRequestedImageEnergyDataNotFound()
```
Comment (lsem) [inline SeamCarve.Back/src/tests/ApiTests/WebApiTests.cs]: Better name for this test is ShouldReturn404ForGettingEnergyOfUnexistingImage.

File: SeamCarve.Back/src/tests/packages.config [hunk 10]
```diff
@@ -1,10 +1,12 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Microsoft.AspNet.Cors" version="5.2.3" targetFramework="net45" />
   <package id="Microsoft.AspNet.WebApi.Client" version="5.2.3" targetFramework="net45" />
   <package id="Microsoft.AspNet.WebApi.Core" version="5.2.3" targetFramework="net45" />
   <package id="Microsoft.AspNet.WebApi.Owin" version="5.2.3" targetFramework="net45" />
   <package id="Microsoft.AspNet.WebApi.OwinSelfHost" version="5.2.3" targetFramework="net45" />
-  <package id="Microsoft.Owin" version="2.0.2" targetFramework="net45" />
+  <package id="Microsoft.Owin" version="3.0.1" targetFramework="net45" />
```
Comment (lsem) [inline SeamCarve.Back/src/tests/packages.config]: Updating any versions of libraries should be performed in separate commit after agreement. We should be careful with this and not to commit what we were not going to commit.

File: SeamCarve.Back/src/Web/App.config [hunk 11]
```diff
@@ -13,6 +13,10 @@
         <assemblyIdentity name="Microsoft.Owin" publicKeyToken="31bf3856ad364e35" culture="neutral" />
         <bindingRedirect oldVersion="0.0.0.0-3.0.1.0" newVersion="3.0.1.0" />
       </dependentAssembly>
+      <dependentAssembly>
```
Comment (lsem) [inline SeamCarve.Back/src/Web/App.config]: the same as with owens, sqlite, etc.. we should have a reason for doing such upgrades.

File: SeamCarve.Back/src/ImageProcesing/Filters/SobelEdgeDetector.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            int outWidth = inMask.Width - 2,
+                outHeight = inMask.Height - 2;
+            var outSize = outWidth * outHeight;
+
+            var outMask = new BmpMask(new byte[outSize], outWidth, outHeight);
+
+            int inWidth = inMask.Width,
+                inHeight = inMask.Height;
+
+            Func<int, int, byte> i = (x, y) => inMask.Data[ y * inWidth + x ];
```
Comment (lsem) [inline SeamCarve.Back/src/ImageProcesing/Filters/SobelEdgeDetector.cs]: please do not use outMask here but array allocated in this function as it was in originally suggested code snippet.
It should be not equivalent from performance perspective.
